nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py

```diff
@@ -209,6 +245,89 @@ def debug(self, msg):
         self.get_logger().debug(msg)
         return
 
+    def getMap(self):
+        request = GetMap.Request()
+        if self.map_client.wait_for_service(timeout_sec=5.0):
+            # Request to get map
+            future = self.map_client.call_async(request)
+            rclpy.spin_until_future_complete(self, future)
+            if future.result() is not None:
+                self.map = future.result().map
+                self.get_logger().info('Laser scan will be populated using map data')
+            else:
+                self.get_logger().warn(
+                    'Failed to get map, '
+                    'Laser scan will be populated using max range'
+                )
+        else:
+            self.get_logger().warn(
+                'Failed to get map, '
+                'Laser scan will be populated using max range'
+            )
+
+    def getLaserPose(self):
+        mat_map_to_odom = transformStampedToMatrix(self.t_map_to_odom)
+        mat_odom_to_base = transformStampedToMatrix(self.t_odom_to_base_link)
+
+        mat_map_to_laser = tf_transformations.concatenate_matrices(
+            mat_map_to_odom,
+            mat_odom_to_base,
+            self.mat_base_to_laser
+        )
+        transform = matrixToTransform(mat_map_to_laser)
+
+        x = transform.translation.x
+        y = transform.translation.y
+        theta = tf_transformations.euler_from_quaternion([
+            transform.rotation.x,
+            transform.rotation.y,
+            transform.rotation.z,
+            transform.rotation.w
+        ])[2]
+
+        return x, y, theta
+
+    def getLaserScan(self, num_samples):
+        if self.map is None or self.initial_pose is None or self.mat_base_to_laser is None:
+            self.scan_msg.ranges = [self.scan_msg.range_max - 0.1] * num_samples
+            return
+
+        x0, y0, theta = self.getLaserPose()
+
+        mx0, my0 = worldToMap(x0, y0, self.map)
+
+        if not 0 < mx0 < self.map.info.width or not 0 < my0 < self.map.info.height:
+            # outside map
+            self.scan_msg.ranges = [self.scan_msg.range_max - 0.1] * num_samples
+            return
+
+        for i in range(num_samples):
+            curr_angle = theta + self.scan_msg.angle_min + i * self.scan_msg.angle_increment
+            x1 = x0 + self.scan_msg.range_max * math.cos(curr_angle)
```

Comment:
`range_max` should be in grid coordinates ( divided by resolution)

Reply:
I don't get it x0 and x1 are world coordinates so basically distance between (x0, y0) and (x1, y1) should be range_max, which should be in meters

let me know if I'm wrong here